### PR TITLE
proto/charybdis: Add support for MECHLIST

### PR DESCRIPTION
**Pull request by Diftraku**
Adds support for MECHLIST in protocol/charybdis

### CODE REVIEW

File: modules/protocol/charybdis.cpp [hunk 1]
```diff
@@ -63,6 +63,18 @@ class CharybdisProto : public IRCDProto
 	void SendLogin(User *u, NickAlias *na) anope_override { ratbox->SendLogin(u, na); }
 	void SendLogout(User *u) anope_override { ratbox->SendLogout(u); }
 
+	void SendSASLMechanisms(std::vector<Anope::string> &mechanisms) anope_override
+	{
+		unsigned i;
```
Comment (SaberUK) [inline modules/protocol/charybdis.cpp]: Declare `i` inside the for loop.

File: modules/protocol/charybdis.cpp [hunk 2]
```diff
@@ -63,6 +63,18 @@ class CharybdisProto : public IRCDProto
...
+		Anope::string mechlist;
+		
+		for (i = 0; i < mechanisms.size(); ++i) {
```
Comment (SaberUK) [inline modules/protocol/charybdis.cpp]: Either put the { on a new line or preferably just omit the {} altogether.

**Comment by Diftraku** [comment]
Updated pull request accordingly, let me know if you want me to squash the style changes into the first commit.

**Comment by kaniini** [comment]
For what it's worth, the reason why people are trying to submit this pull request is because an Anope contributor (@SaberUK) specifically requested that we provide a patch to add support for charybdis to supplement his previous work.  So we did.

```
16:48 --> SaberUK (~saberuk@46.166.188.203) has joined #charybdis
16:52 <SaberUK> FYI: anope now has support for sending sasl mechanism lists to the ircd for use in the sasl-3.2 cap value, i've only implemented it for inspircd though because i dont know about other ircd protocols so if you use anope you might want to implement it in the charybdis module.
16:52 <SaberUK> heres the inspircd implementation for reference. https://github.com/anope/anope/blob/52fdc455976e22e860736b07c431c6a6464f5d70/modules/protocol/inspircd20.cpp#L45-L52
17:05 <@kaniini> i'll look into it this weekend
17:05 <SaberUK> :)
```

This isn't some massive plan to harm Anope, we are just trying to ensure the 17 networks running Charybdis + Anope have feature parity with Charybdis + Atheme.  This is called being a responsible IRCd vendor.

**Comment by awilfox** [comment]
I'm going to be honest here, I tried to set up Anope with Charybdis for my IRC test farm, where I use different servers and services packages in a mix and match fashion to test conformance and client behaviour.  I had issues with SASL that would be resolved with this, having just applied the patch to that test network.

Please don't let personal biases affect the quality and feature set of Anope.

**Comment by Adam-** [comment]
As I stated previously, if someone else re-implemented it, tested it, and submitted it, I would merge it.
I am certainly not letting personal biases affect Anope, I think @awilfox has been drinking the kool aid a little too much.

I am responsible for defending Anope, and in the past people affiliated with other projects have used their code contributions to
harm the project by using it to spreading propaganda. As such, I become concerned when users like yourselves, who don't use Anope,
who hate it and would harm it if given the opportunity, start sending pull requests to forward your own projects/agenda.

Let's not fool anyone here. You don't use Anope, and noone who uses Charybdis uses Anope. I had the support for Charybdis added for one
network specifically which has since moved to InspIRCd. As such, I have *no* problem removing all Charybdis support entirely if this gets
escalated further in the form of legal threats or otherwise.

I find completely untested pull requests a waste of time. If I have to go test them I may as well write it myself. I have seen
many contributions/modules from @kaniini to various projects that do not compile, including previous Anope modules, because he refuses to test them
because the build system is cmake. He is also impossible to reason, which is one of the reasons I have closed his request.

I see @awilfox has tested it, which I will assume is true, so thanks. @Diftraku please squash this and I will merge it.

**Comment by Diftraku** [comment]
@Adam- squash'd

**Comment by GLolol** [comment]
> I am responsible for defending Anope, and in the past people affiliated with other projects have used their code contributions to
harm the project by using it to spreading propaganda. As such, I become concerned when users like yourselves, who don't use Anope,
who hate it and would harm it if given the opportunity, start sending pull requests to forward your own projects/agenda.

I think it's about time everyone just puts their history away and look at code based on the lines of work they represent, and not who wrote it. Surely everyone has their personal agenda; it's the only way to get *anything* done. Someone who truly hates a project would probably not even bother to send patches in the first place.

> As such, I have no problem removing all Charybdis support entirely if this gets
escalated further in the form of legal threats or otherwise.

How is this even close to a legal threat?

----
*disclaimer*: I'm not affiliated with either Anope, Atheme, or any IRCd vendor. This is just my personal thoughts from seeing this issue so far.

**Comment by awilfox** [comment]
> I think @awilfox has been drinking the kool aid a little too much.

From .weechat/logs/irc.chatspike.#inspircd.weechatlog:

```
2015-08-04 15:04:45     TheWilfox       you're going to want to look up either anope or shalture, those are the two main services packages
```
```
2015-09-29 15:35:32     TheWilfox       anope is easier to configure
2015-09-29 15:39:42     TheWilfox       performance is a tossup; both will take you to 25k users
```

If this "kool aid" has anything to do with my contributions to Charybdis, I'd just like to point to [the Gentoo metadata for InspIRCd](https://github.com/gentoo/gentoo/blob/master/net-irc/inspircd/metadata.xml), which lists me as maintainer.  I contribute when I can, to any package I can, because I believe that egalitarian contributions will enrich the entire IRC ecosystem.

> I become concerned when users like yourselves, who don't use Anope, who hate it and would harm it if given the opportunity

Right, okay.  I _just_ twelve hours ago said I use Anope on my test net, and I have in the past assisted people on #inspircd with it when I could because I had experience with it.

> I had the support for Charybdis added for one network specifically which has since moved to InspIRCd. As such, I have _no_ problem removing all Charybdis support entirely if this gets escalated further in the form of legal threats or otherwise.

I don't see any form of escalation or legal threats.  I want to see the IRC community grow together and adopt more compatibility between servers, clients, service packages, bot frameworks, and so on.  I would hope that you share that goal.

**Comment by Elizafox** [comment]
William wanted me to pass on, since apparently he is banned for some reason (not surprising given your attitude of telling others to literally commit suicide, often with detailed instructions):

> > I am responsible for defending Anope, and in the past people affiliated with other projects have used their code contributions to harm the project by using it to spreading propaganda. As such, I become concerned when users like yourselves, who don't use Anope, who hate it and would harm it if given the opportunity, start sending pull requests to forward your own projects/agenda.
>
> That seems like a complete and total misunderstanding of the situation.  Atheme's only issue with Anope (oh, did you know I don't even have any formal affiliation with the Atheme group anymore) was the misappropriation of `include/patricia.h` into Anope without copyright notices, with your justification of this being "I ported it to C++".  Beyond that, and even before that, nobody cared and we actually collaborated occasionally on some issues.  I am just bringing this up because I do not understand your stance; in a personal review of my own IRC logs from 2013 (excluding 2012 because that is when the `include/patricia.h` problem occured) to present for example, I have only used 'anope' in a sentence a few dozen times, and have actually recommended anope to people who did not have a good experience with atheme.
>
> My agenda is to ensure that charybdis users who use anope get to take advantage of the same features and functionality as those charybdis users who use atheme or a derivative.  This seems like a good agenda for advancing mutual interests -- specifically advancing the available features to Charybdis + Anope networks.
>
> > Let's not fool anyone here. You don't use Anope, and noone who uses Charybdis uses Anope. I had the support for Charybdis added for one network specifically which has since moved to InspIRCd.
>
> Actually, on the charybdis testnet we were running Anope for some time.  We switch between them whimsically.
>
> Beyond that, we occasionally get people in #charybdis asking about using it with Anope.  It is a fully supported configuration on our end, as well.  As previously mentioned, there are at least 17 networks that I know of running Anope + Charybdis.
>
> As such, I have no problem removing all Charybdis support entirely if this gets escalated further in the form of legal threats or otherwise.
>
> We sent you a pull request and explained that we were doing so for the benefit of our mutual users, to which you refused to merge it and explained that you wanted a third party to resubmit it. So they did, because the charybdis community (especially those who run Charybdis + Anope), are interested in having this functionality. I think removing the support because we sent you a pull request to improve the support is only harmful to Anope, as it will leave users using your software without an upgrade path. At any rate, if it were to happen we would probably provide an Anope tree with the support patched back in.

**Comment by Adam-** [comment]
@awilfox "kool aid" as in you think I am putting personal bias into this. If I was, I would remove charybdis support, sasl support, and ban every single one of you. But, as I have stated several times already I am willing to add support for this.

@GLolol That's not quite how it works. Let's hope you never have to understand the type of shit I've had to go through over the years due to kaniini and his various minions.

@Elizafox I don't know who you are, but I have never told anyone to commit suicide, and certainly not with any detailed instructions. It sounds to me like you have been brainwashed by those from #atheme. This is why I try to stay away from you guys (character assassinations, lies, atheme propaganda etc.). Anyone who knows me know I would never do anything like that, and I'm sorry you think that.

I'll merge this now, please, all of you, just leave me alone.

**Pull request merged**